examples/chip-tool/commands/common/BDXDiagnosticLogsServerDelegate.cpp

```diff
@@ -0,0 +1,192 @@
+{
+    VerifyOrReturnError(fileDesignator.size() <= kMaxFileDesignatorLen, CHIP_ERROR_INVALID_STRING_LENGTH);
+
+    memcpy(outFilePath.data(), kTmpDir, sizeof(kTmpDir));
+    memcpy(outFilePath.data() + sizeof(kTmpDir) - 1, fileDesignator.data(), fileDesignator.size());
+    outFilePath.reduce_size(sizeof(kTmpDir) + fileDesignator.size());
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CheckFileExists(const char * filePath)
+{
+    if (access(filePath, F_OK) != 0)
+    {
+        ChipLogError(chipTool, "The file '%s' for dumping the logs does not exist.", filePath);
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CheckFileDoesNotExist(const char * filePath)
+{
+    if (access(filePath, F_OK) == 0)
+    {
+        ChipLogError(chipTool, "The file '%s' for dumping the logs already exists.", filePath);
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void LogFileDesignator(const char * prefix, const chip::CharSpan & fileDesignator, CHIP_ERROR error = CHIP_NO_ERROR)
+{
+#if CHIP_PROGRESS_LOGGING
+    auto size = static_cast<uint16_t>(fileDesignator.size());
+    auto data = fileDesignator.data();
+    ChipLogProgress(chipTool, "%s (%u): %.*s", prefix, size, size, data);
+#endif // CHIP_PROGRESS_LOGGING
+
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(chipTool, "%s: %s", prefix, ErrorStr(error));
+    }
+}
+
+CHIP_ERROR CreateFile(const char * filePath)
+{
+    VerifyOrReturnError(nullptr != filePath, CHIP_ERROR_INVALID_ARGUMENT);
+
+    auto fd = fopen(filePath, "w+");
+    VerifyOrReturnError(nullptr != fd, CHIP_ERROR_WRITE_FAILED);
+
+    auto rv = fclose(fd);
+    VerifyOrReturnError(EOF != rv, CHIP_ERROR_INTERNAL);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AppendToFile(const char * filePath, const chip::ByteSpan & data)
+{
+    VerifyOrReturnError(nullptr != filePath, CHIP_ERROR_INVALID_ARGUMENT);
+
+    auto fd = fopen(filePath, "a");
+    VerifyOrReturnError(nullptr != fd, CHIP_ERROR_WRITE_FAILED);
+
+    fwrite(data.data(), data.size(), 1, fd);
+
+    auto rv = fclose(fd);
+    VerifyOrReturnError(EOF != rv, CHIP_ERROR_INTERNAL);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BDXDiagnosticLogsServerDelegate::OnTransferBegin(chip::bdx::BDXTransferProxy * transfer)
+{
+    auto fileDesignator = transfer->GetFileDesignator();
+    LogFileDesignator("OnTransferBegin", fileDesignator);
+
+    chip::CharSpan phaseErrorTarget(kErrorOnTransferBegin, sizeof(kErrorOnTransferBegin) - 1);
+    ReturnErrorOnFailure(CheckForErrorRequested(phaseErrorTarget, fileDesignator));
+    ReturnErrorOnFailure(CheckFileDesignatorAllowed(mFileDesignators, fileDesignator));
+
+    char outputFilePath[kMaxFilePathLen] = { 0 };
+    chip::MutableCharSpan outputFilePathSpan(outputFilePath);
+    ReturnErrorOnFailure(GetFilePath(fileDesignator, outputFilePathSpan));
+    ReturnErrorOnFailure(CheckFileDoesNotExist(outputFilePath));
```

Comment:
```suggestion
    // Ensure null-termination of the filename, since we will be passing it around as a char *.
    VerifyOrReturnError(outputFilePathSpan.size() < ArraySize(outputFilePath), CHIP_ERROR_INTERNAL);
    outputFilePath[outputFilePathSpan.size()] = '\0';
    ReturnErrorOnFailure(CheckFileDoesNotExist(outputFilePath));
```